app/javascript/types/account.ts

```diff
@@ -1,27 +1,31 @@
 import type { Record } from 'immutable';
 
-type CustomEmoji = Record<{
-  shortcode: string;
-  static_url: string;
-  url: string;
-}>;
+import type { CustomEmoji, CustomEmojiRawValues } from './custom_emoji';
 
-type AccountField = Record<{
+interface AccountFieldRawValues {
   name: string;
   value: string;
   verified_at: string | null;
-}>;
+}
+type AccountField = Record<AccountFieldRawValues>;
+
+interface AccountRoleRawValues {
+  color: string;
+  id: string;
+  name: string;
+}
+type AccountRole = Record<AccountRoleRawValues>;
```

Comment:
Same comment here as [in your other PR](https://github.com/mastodon/mastodon/pull/25521/files#r1242654254): are those really `Record`, and not `Map`?

Note: I would really prefer them to be `Record`, but I am not sure they really are.
You can `console.log` them to confirm it.

Reply:
Yes, this is actually a Map, not a Record.
However, the type annotation of Immutable Map returns union type.
This is useless for our usage. This is just trush, in this case.

```typescript
import { Map, Record } from 'immutable';

const someMap = Map({ foo: 'foo value', bar: 2 });
const mapFoo = someMap.get('foo');

const someRecord = Record({ foo: 'foo value', bar: 2 })();
const recordFoo = someRecord.get('foo');

interface MastodonMap<T> {
  get<K extends keyof T>(key: K): T[K];
  has<K extends keyof T>(key: K): boolean;
  set<K extends keyof T>(key: K, value: T[K]): this;
}

const someMastodonMap: MastodonMap<{ foo: string; bar: number }> = Map({
  foo: 'foo value',
  bar: 2,
});
const mastodonMapFoo = someRecord.get('foo');
```
![image](https://github.com/mastodon/mastodon/assets/24884114/3c5218d5-5eb8-427e-a809-4b7d254256ac)
![image](https://github.com/mastodon/mastodon/assets/24884114/5996fa8f-8647-41c2-b775-1fc0e1c845c7)
![image](https://github.com/mastodon/mastodon/assets/24884114/4da4ce62-550c-4552-bf89-6cdaf9515e74)

Do we revive MastodonMap?

Reply:
The issue with using `MastodonMap` is that you will need to redefine every other method that is defined on `Map`, and it hides the fact that some keys can be missing in it, or the underlying types can be incorrect.
This is fine for now as we dont do much with the state in Typescript for now, but will be an issue once we start to convert components using the state into Typescript.

I am not sure it is a good idea to define them as `Record` when they are not as incorrect types are really not a good idea. But I also know that defining those as `ImmutableMap` is not really useful, but is strictly correct.

I need to do some research to see how hard it would be to convert those into proper `Record` when they are created from the server.